### ISSUE TITLE
Custom Selector Feature

### ISSUE THREAD

**Issue by ardalis** [ardalis/Specification#2]
Would like to be able to add custom select to the specification.

**Comment by fiseni** [comment]
Hey @ardalis,

Do you need any help/contribution to this project? I liked how this pattern is been used in eShopOnWeb, and I have tendency to use it in my own projects. I have been using my own specification package. I see you already have nuget package for this, and I would happily contribute. It would be nice to have one consolidated package.

As for the particular issue with the selector feature, I have the exact construct.
```
protected virtual void AddSelect(Expression<Func<T, object>> select)
{
	Select = select;
}
public Expression<Func<T, object>> Select { get; private set; }
```
But, in day to day use, it turned out that having "object" as an output is a bit troublesome. Casting the results all the time, leads to not type-safe code. Instead, I think you get better usage if the feature is part of the repository as shown below, and not part of the specification.
```
public async Task<IReadOnlyList<TResult>> ListSelectAsync<TResult>(ISpecification<T> specification, Expression<Func<T, TResult>> selector)
{
	return await ApplySpecification(specification).Select(selector).ToListAsync();
}
```

**Comment by ardalis** [comment]
Certainly open to help and ideas.

**Comment by fiseni** [comment]
The selector functionality, as it is done here, probably won't work. The select always returns new type, so the following will fail (it wont be able to convert object to T)
```
// Apply selector
query = query.Select(specification.Selector);
```
I created PR #5 in this branch. If you still want the selector within the specification, something like this might do the trick. I tried to keep selector evaluation within specification logic, and not delegating to the repository.

**Comment by ardalis** [comment]
I'll try and look at your PR during my stream tomorrow, 12-2 EST. Feel free to hop on twitch.tv/ardalis if you're around then and we can discuss in real time.

**Comment by fiseni** [comment]
Thanks @ardalis. I just checked your channel and watched your previous video. Great content. Probably I won't make it for today, timezones :)

I just have a small note. You couldn't use ListAsync, because you were working upon IEnumerable type. The reason is that you were using `Func<T, object>` for the selector property, instead of `Expression<Func<T, object>>`.
If you use `Func<T, object>` on IQueryable, it would return IEnumerable. You need IQueryable, so you can continue with building the query and not materializing it at that moment. Expression<> would do the trick (it returns IQueryable). Then you can apply ToListAsync for sure.

**Comment by ardalis** [comment]
Fixed by #4 and #5
